src/cargo/util/semver_ext.rs

```diff
@@ -84,10 +84,17 @@ impl OptVersionReq {
             OptVersionReq::Any => true,
             OptVersionReq::Req(req) => req.matches(version),
             OptVersionReq::Locked(v, _) => {
```

Comment:
There are three semantics used for two versions being equal.
1. constructing a `=` requirement from `A` and seeing if the requirement matches `B`. This is what the old code does, for backwards compatibility with before we had `Locked`. (cc @weihanglo re https://github.com/rust-lang/cargo/pull/9847#discussion_r697887365)
2. directly checking if 'A == B`, requiring all fields to match. The comet argues that this should be correct, because we only set this to a value of a version from the lock file which should be fully specified version. All tests pass if we change to that behavior. but tests don't always consider all the weird things in the world like versions that only different metadata.
3. The comparison used for `--precise` witch allows the metadata fields to be anything if the past in version does not have a metadata field.

In this PR I make `Locked` more strict then 1, but only to the semantics of 3. This simplifies the handling of `--precise`. but maybe conflating unrelated issues. Perhaps it would be better to have two different variants for the two different semantics.

Reply:
Can/should this change be a separate commit?

Also, why do we check `metadata` at all?  I thought it isn't "significant" in versions and in various places its left off and I could see problems with making sure we deal with mixed cases correctly (like here)

Reply:
I had a lot to say in response to this, and decided it needed its own PR #12772.